source/Conversation.cpp

```diff
@@ -434,46 +452,122 @@ const Sprite *Conversation::Scene(int node) const
 
 
 // Find out where the conversation goes if the given option is chosen.
-int Conversation::NextNode(int node, int choice) const
+int Conversation::NextNode(int node, int element) const
 {
-	if(static_cast<unsigned>(node) >= nodes.size()
-			|| static_cast<unsigned>(choice) >= nodes[node].data.size())
+	if(!NodeIsValid(node) || !ElementIsValid(node, element))
 		return DECLINE;
 
-	return nodes[node].data[choice].second;
+	return nodes[node].elements[element].next;
+}
+
+
+
+// Returns whether the given node should be displayed.
+bool Conversation::ShouldShowText(const map<string, int64_t> &vars, int node, int element) const
+{
+	if(!NodeIsValid(node))
+		return false;
+	else if(IsChoice(node) ? !ElementIsValid(node, element) : element != 0)
+		return false;
+	const auto &data = nodes[node].elements[element];
+	if(data.conditions.IsEmpty())
+		return true;
+	else
+		return data.conditions.Test(vars);
 }
 
 
 
-// Parse the children of the given node to see if then contain any "gotos."
-// If so, link them up properly. Return true if gotos were found.
+// Returns true if the given node index is in the range of valid nodes for this
+// Conversation.
+bool Conversation::NodeIsValid(int node) const
+{
+	if(node < 0)
+		return false;
+	else
+		return static_cast<unsigned>(node) < nodes.size();
+}
+
+
+
+// Returns true if the given node index is in the range of valid nodes for this
+// Conversation *and* the given element index is in the range of valid elements
+// for the given node.
+bool Conversation::ElementIsValid(int node, int element) const
+{
+	if(!NodeIsValid(node))
+		return false;
+	else if(element < 0)
+		return false;
+	else
+		return static_cast<unsigned>(element) < nodes[node].elements.size();
+}
+
+
+
+// Parse the children of the given node to see if then contain any "gotos," or
+// "to shows." If so, link them up properly. Return true if gotos or
+// conditions were found.
 bool Conversation::LoadGotos(const DataNode &node)
 {
 	bool hasGoto = false;
+	bool hasCondition = false;
 	for(const DataNode &child : node)
 	{
-		if(hasGoto)
+		if(hasGoto && hasCondition)
 			child.PrintTrace("Warning: Ignoring extra text in conversation choice:");
 		else if(child.Size() == 2 && child.Token(0) == "goto")
 		{
 			// Each choice can only have one goto
-			Goto(child.Token(1), nodes.size() - 1, nodes.back().data.size() - 1);
-			hasGoto = true;
+			if(hasGoto)
+				child.PrintTrace("Warning: Ignoring extra endpoint in conversation choice:");
+			else
+			{
+				Goto(child.Token(1), nodes.size() - 1, nodes.back().elements.size() - 1);
+				hasGoto = true;
+			}
+		}
+		else if(child.Size() == 2 && child.Token(0) == "to" && child.Token(1) == "show")
```

Comment:
How about "display" instead of "show."

Reply:
That may be better. I'll think about it some more.

Reply:
I was leaning towards `to display`, but then I found that there is currently one other use of `to show`: News. Should we mirror that usage, or is it better to have a different phrasing here?

Reply:
Well there's no other `to *` phrase that's duplicated anywhere else that I can recall, hence why I suggested the alternative of `display`.

Reply:
Just ran a quick search:

- `to show` in News
- `to spawn`/`despawn` in NPC
- `to` in Interface (doesn't count)
- `to offer`/`complete`/`fail` in Mission

There are indeed no existing pairs. Since every `to *` phrase is currently unique, I'll go ahead and make this one `to display`.